### DISCUSSION TITLE
Move towards changelog generation

### DISCUSSION THREAD

**Discussion by Darhazer** [rubocop-hq/rubocop-rspec#1417] [General]
In the past months I notice much higher activity in the rubocop-rspec. This means that after any merge, a lot of PRs have to be updated to resolve the changelog conflict, just to be updated again on the next merge. Perhaps it's time to move to the same strategy of changelog being generated, with individual changelog files for each PR, like in rubocop and rubocop-rails ?

**Comment by pirj** [comment]
I'm just wondering if there's an easier way than the changelog/ folder.
Like, e.g instead of putting the entry below other, put it on top of the file?

What I did:

1. created a file on `main`:
```

* two


1.1

* foo
* bar
```

2. branched off `one` and `two`, each with a commit adding one line:

```diff

+* one

1.1

* foo
* bar
```

```diff

+* two

1.1

* foo
* bar
```

3. merged `one` to `main`
4. created a release commit on `main`:
```diff

+
+ 1.2

* one

1.1

* foo
* bar
```

5. Rebased `two` from `main`:
```


* two

1.2

* one

1.1

* foo
* bar
```

🎉 
cc @bquorning 

Should we ask someone from `@rubocop/rubocop-core` if they've considered this?
cc @koic WDYT?

**Reply by bquorning** [reply]
What happens if you skip step 4 above? Most conflicts happen _between_ releases, although I am aware that changelog entries often end up in the wrong section _after_ a release too.

**Reply by pirj** [reply]
This, indeed, results in a merge conflict 😢
```
<<<<<<< HEAD
* three
||||||| cc31063
=======
* two
>>>>>>> two

1.2
```

**Comment by pirj** [comment]
[Using a per-file custom git merge strategy](https://stackoverflow.com/questions/22109586/what-is-the-difference-between-union-and-text/22109606#22109606) `.gitattributes`:
```
CHANGELOG.md      text merge=union
```

```
main $ git merge two
```

```
* three
* two

1.2

* one

1.1

* foo
* bar
```

🎉

**Reply by pirj** [reply]
Well, we kind of already had that:
```
75fc8d38d (John Backus 2017-01-12 22:10:12 -0800> CHANGELOG.md merge=union
```

but it missed the `text` part.
I've sent [a PR](https://github.com/rubocop/rubocop-rspec/pull/1425).

Let's stick to the strategy of:
1. removing the header
```
# Changelog

## Master (Unreleased)
```
2. Asking contributors to add their change to the top
3. Observe how it goes

**Reply by pirj** [reply]
Note to self - adding the changelog entry to the top (e.g. https://github.com/rubocop/rubocop-rspec/pull/1438) still caused other PRs to have a merge conflict.

**Comment by bquorning** [comment]
I tried the `CHANGELOG.md text merge=union` in a private repo. Merging locally works fine, but GitHub still sees a conflict.

I searched a little bit around and found these interesting posts:
- https://github.com/orgs/community/discussions/9288
- https://github.com/isaacs/github/issues/487
- https://github.com/olivierlacan/keep-a-changelog/issues/56

tl;dr: Github UI still doesn’t support `merge=union`, but it seems that GitLab does.

**Reply by pirj** [reply]
Some say there's an [alternative manual approach](https://stackoverflow.com/a/34517149/202914) to achieve the same:
```
[merge "aggregate"]
    name = aggregate both new sections
    driver = git merge-file --union -L %P %A %O %B
```
However, `.git/config` is not pushed, and adding this to either `.gitconfig` or `.git/config` didn't help.
And I can't find `merge.aggregate` in `man git-config` 🤔

**Reply by bquorning** [reply]
If GitHub doesn’t respect .gitattributes, and we cannot push any git config to the repository, then I cannot see how we can fix the issue without splitting the changelog entries into separate files (or into separate parts of the same file, which I'm not sure would help at all).

**Reply by pirj** [reply]
GitHub actions could have helped, but ["Workflows will not run on pull_request activity if the pull request has a merge conflict"](https://github.com/orgs/community/discussions/26304).

There's [an action that marks a PR with a label when it needs a rebase](https://github.com/marketplace/actions/label-conflicting-pull-requests) ([source](https://github.com/eps1lon/actions-label-merge-conflict/blob/main/sources/main.ts)). Wondering if it's possible to adjust it to rebase it using a script.

**Reply by bquorning** [reply]
I imagine having your branch rebased by a script would feel disruptive to our contributors.

**Reply by pirj** [reply]
To round it up, we're dealing with two issues here:
1. Maintainers can't merge more than one PR in a row without getting a merge conflict on the subsequent PRs
2. Contributors or maintainers have to rebase and force-push before the PR can be merged

On the other hand, the file-based changelog entry approach:
1. (subjective) Offers less discoverability into what has been added in the integration branch
2. (?) adds a one command step to the release process
3. + offers solutions for the two main issues we have
4. (subjective) adds some cognitive overhead to entry file creation

Given there are no sane options to deal with a single changelog file, I give up and agree with the multi-file appoach.

**Comment by koic** [comment]
I'm late riding on this one.
If you like, I'll bring RuboCop and RuboCop Rails changelog logic to RuboCop RSpec. 

- https://github.com/rubocop/rubocop-rails/pull/496 and other PRs

After introducing the current RuboCop changelog logic, the number of collisions has definitely decreased.
OTOH, It differs from @pirj approach #1425 which has already been opened. If you don't mind, shall I open a PR to bring the RuboCop changelog logic? (It may be a little later because I'm busy with other work on this week)

Or, if you try #1425 first, I'll pend my action.

**Reply by pirj** [reply]
This is very kind of you, @koic . I've closed #1425 in favour of a commonly used approach.

**Reply by koic** [reply]
I got it. Please be patient while I open a PR.

**Comment by ydah** [comment]
Once attempted, but currently unnecessary, so will proceed again when demand increases:
- https://github.com/rubocop/rubocop-rspec/pull/1542
